### ISSUE TITLE
Conflicting data from show series and select: select is merging two series with different tags

### ISSUE THREAD

**Issue by mjibson** [influxdb/influxdb#3059]
I'm getting conflicting data from a `SHOW SERIES` and a `SELECT * FROM a`. With a fresh InfluxDB 0.9.0 server, I run the following script to populate some data and then query it:

```
curl -G http://localhost:8086/query --data-urlencode "q=CREATE DATABASE db"
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,b=m value=1 1'
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,c=m value=2 2'
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,c=n value=3 3'
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,b=n value=4 4'

echo SELECT COUNTS
curl -s -G 'http://localhost:8086/query?db=db' --data-urlencode "q=SELECT * FROM a" | jq -S -M ".results[].series | map(.tags) | length"

echo SHOW SERIES COUNTS
curl -s -G 'http://localhost:8086/query?db=db' --data-urlencode "q=SHOW SERIES" | jq -S -M ".results[].series[].values | length"

curl -s -G 'http://localhost:8086/query?db=db' --data-urlencode "q=SELECT * FROM a" | jq -S -M "."
```

This creates the measurement `a` with 4 unique tag sets, values, and times. The result to `SHOW SERIES` says that there are 4 such unique series. Yet the `SELECT` only returns 2 results. The final `curl` command there shows us why. Here's a partial result of it:

```
        {
          "columns": [
            "time",
            "value"
          ],
          "name": "a",
          "tags": {
            "b": "m",
            "c": ""
          },
          "values": [
            [
              "1970-01-01T00:00:00.000000001Z",
              1
            ],
            [
              "1970-01-01T00:00:00.000000002Z",
              2
            ]
          ]
        },
```

Here there are 2 values in this series identified by the tags `b=m,c=` (i.e., c is empty). Yet from the curl insertion above we see that the point with value 2 had the `c` tag set to `m` and `b` to empty.

Is this a bug or do I misunderstand how tags work?

---

As an aside, I found while testing this with a much larger number of unique tag sets (instead of just 4), that InfluxDB will return different results each time an identical query is made. The tags returned change per query. This was occurring for me when `SHOW SERIES` was returning 42 unique series, and the `SELECT` would then return some set of 12 of them, but different every time.

**Comment by beckettsean** [comment]
@mjibson great find! You don't misunderstand tags, this appears to be a bug. It looks like the results are being grouped by tag values without first grouping by tag keys. 

Recreating using CLI:

```
> insert result_grouping,t1=val1 value=1
> insert result_grouping,t2=val2 value=2
> select * from result_grouping
name: result_grouping
tags: t1=val1, t2=
time				value
----				-----
2015-06-19T20:18:53.565339084Z	1

name: result_grouping
tags: t1=, t2=val2
time				value
----				-----
2015-06-19T20:19:01.228395991Z	2
```

For two points with different tag values, we see two tag set groupings.

Add a third point that shares a tag value but makes a new tag set, and it gets incorporated into an already existing tag set group when the results are returned.

```
> insert result_grouping,t1=val2 value=3
> select * from result_grouping
name: result_grouping
tags: t1=val1, t2=
time				value
----				-----
2015-06-19T20:18:53.565339084Z	1


name: result_grouping
tags: t1=val2, t2=
time				value
----				-----
2015-06-19T20:19:01.228395991Z	2
2015-06-19T20:19:18.736019871Z	3
```

Any unique tag values generate new tag set groups in the results. Common tag values for different tag keys, however, all get grouped together in the same tag set, even though they are not equivalent.

**Comment by beckettsean** [comment]
There's a workaround of sorts. The data is still recoverable by using explicit tag queries, e.g. `select * from result_grouping where t1 =~ /.*/`

**Comment by otoolep** [comment]
The problem does not seem to occur when the query selects a specific field.

**Comment by otoolep** [comment]
Though tags aren't included in the result when selecting against a specific field.

**Comment by otoolep** [comment]
The statement `SELECT * FROM a` is actually re-written by the system to `SELECT value FROM a GROUP BY c,b`. This is because the system *defines* `SELECT *` as selecting all fields *and* tags.

There still appears to be a bug here. A full session follows below.

```
curl -G http://localhost:8086/query --data-urlencode "q=CREATE DATABASE db"
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,b=m value=1 1'
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,c=m value=2 2'
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,c=n value=3 3'
curl -XPOST 'http://localhost:8086/write?db=db' -d 'a,b=n value=4 4'
curl -s -G 'http://localhost:8086/query?db=db&pretty=true' --data-urlencode "q=SELECT value FROM a GROUP BY c,b"
{
    "results": [
        {
            "series": [
                {
                    "name": "a",
                    "tags": {
                        "b": "m",
                        "c": ""
                    },
                    "columns": [
                        "time",
                        "value"
                    ],
                    "values": [
                        [
                            "1970-01-01T00:00:00.000000001Z",
                            1
                        ],
                        [
                            "1970-01-01T00:00:00.000000002Z",
                            2
                        ]
                    ]
                },
                {
                    "name": "a",
                    "tags": {
                        "b": "",
                        "c": "n"
                    },
                    "columns": [
                        "time",
                        "value"
                    ],
                    "values": [
                        [
                            "1970-01-01T00:00:00.000000003Z",
                            3
                        ],
                        [
                            "1970-01-01T00:00:00.000000004Z",
                            4
                        ]
                    ]
                }
            ]
        }
    ]
}
```
As can be seen above the point with a `value` of 2 has a tag `c=m`, but has been grouped with the block `c:""`.

**Comment by otoolep** [comment]
Some unit test to learn more about this issue: https://github.com/influxdb/influxdb/pull/3068

**Comment by otoolep** [comment]
I believe this https://github.com/influxdb/influxdb/pull/3075 will address this issue.

**Comment by otoolep** [comment]
All fixed in 0.9.1.

Thanks for the excellent bug report @mjibson.
